[Thumb] Resolve FIXME: Transform "(and (shl x, c2), c1)" into "(shl (and x, c1>>c2), c2)"

AZero13 opened the pull request:
Transform "(and (shl x, c2), c1)" into "(shl (and x, c1>>c2), c2)" if "c1 >> c2" is a cheaper immediate than "c1" using HasLowerConstantMaterializationCost.

AtariDreams commented:
@topperc Thoughts?

AtariDreams commented:
@topperc Thoughts?

TNorthover commented:
I think this one looks reasonable to me. Logic seems to check out and it looks for benefits before proceeding.

AtariDreams commented:
@RKSimon what do you think about merging this?

AZero13 committed:
[Thumb] Resolve FIXME: Transform "(and (shl x, c2), c1)" into "(shl (and x, c1>>c2), c2)"

Transform "(and (shl x, c2), c1)" into "(shl (and x, c1>>c2), c2)" if "c1 >> c2" is a cheaper immediate than "c1" using HasLowerConstantMaterializationCost.

RKSimon commented:
> @RKSimon what do you think about merging this?

I don't think its necessary - we're up top 18.1.6 now and this is a perf improvement, not a critical bugfix for a recent issue. 

Better to focus on getting trunk ready for 19.X in a couple of months time.

AtariDreams commented:
This wasn't for 18.x

RKSimon commented:
Sorry, misunderstood your question
